master/pkg/stream/stream.go

```diff
@@ -2,22 +2,55 @@ package stream
 
 import (
 	"sync"
-
-	"github.com/gorilla/websocket"
 )
 
+// UpsertFunc is a function that overrides the default upsert
+type UpsertFunc func(interface{}) interface{}
+
+// DeleteFunc is a function that overrides the default deletion
+type DeleteFunc func(string, string) interface{}
+
 // Msg is an object with a message and a sequence number and json marshal caching.
 type Msg interface {
 	SeqNum() int64
-	UpsertMsg() *websocket.PreparedMessage
-	DeleteMsg() *websocket.PreparedMessage
+	//PrepareFn(upsert UpsertMsg, delete DeleteMsg) interface{}
+	//UpsertMsg(prepare UpsertFunc) interface{} // cache aversion of yourself for the purpose of upsert on the wire. This would have to take this prep function and pass the prep function to the upsert pmsg fn
+	//// publisher stream in internal, pass each publisher object the prep function. Know it works if you don't have to import websockets into the pkg stream object. When testing, just write in whatever is most useful for the test. Could be the full object.
+	//DeleteMsg(prepare DeleteFunc) interface{}
 }
 
 // Event contains the old and new version a Msg.  Inserts will have Before==nil, deletions will
 // have After==nil.
 type Event[T Msg] struct {
 	Before *T `json:"before"`
-	After *T `json:"after"`
+	After  *T `json:"after"`
+
+	upsertCache interface{}
+	deleteCache interface{}
+}
+
+type UpsertMsg struct {
+	JsonKey string
+	Msg     Msg
+}
+
+func NewUpsertMsg(msg Msg, jsonKey string) UpsertMsg {
+	return UpsertMsg{
+		JsonKey: jsonKey,
+		Msg:     msg,
+	}
+}
```

Comment:
Not sure if we have a code style guide that mandates a constructor for every object, but this strikes me as not really very helpful.

Reply:
agreed...looking again this is very useless